Library: re-focus tracks table when scrolling with controller

ronso0 opened the pull request:
A first step to solve [lp:1673196 Controls for Library focus](https://bugs.launchpad.net/mixxx/+bug/1673196)

When currently no library widget has focus this ensures tracks table has focus after issueing
`[Library],Move/MoveUp/MoveDown`
`[Library],Scroll/ScrollUp/ScrollDown`
`[Library],MoveFocus/~Forward/~Backward`
This also allows to jump directly to the tracks table from the search box.

~~I also added some temp debug output to maybe also focus the lib table directly instead of first focusing the sidebar and sending Tab from there. Maybe I can do this with your help, otherwise I'll just revert that commit.~~

**EDIT**
it turned out emitKeyEvent would already take care of re-focussing the tracks table.
I rolled back my attempt and fixed that function instead.


Direct focus COs would handy for the multi-pane lib-redesign which store the previously focused lib widget.

ronso0 commented on src/library/librarycontrol.cpp:
```diff
@@ -378,23 +387,38 @@ void LibraryControl::slotMoveFocusForward(double v) {
+            qDebug() << " ";
+            qDebug() << " 			  if (!focusWidget) {";
+            qDebug() << " 			    return; }";
+            qDebug() << " ";
```
temp debug output. blank lines and indentations so it's easier to spot in the log

uklotzde replied:
Qt5 has qTrace() for this purpose. You may also use our simple `kLogger` pattern that adds context to all log lines.

ronso0 replied:
@uklotzde I don't find any info on qtrace(), can you point me to some ressources?
Also I don't find a place where `kLogger` usage is explained.

uklotzde commented:
Put the following code into the .cpp file:

```cpp
#include "util/logger.h"

namespace {

const Logger kLogger("LibraryControl");

} // anonymous namespace
```

Use this for logging:
```cpp
kLogger.trace() << "A trace log message";
```

Unfortunately Qt5 does not provide a *trace* level, I was wrong. Currently *trace* logs are logged with level *debug*, independent of the log level setting. I will fix the `Logger` class to suppress trace logs if not explicitly enabled.

uklotzde commented:
Suppressing trace logs would require to return a dummy QDebug object. Not worth to complicate the current implementation.

Please enclose your trace log outputs in log guards manually:
```
if (kLogger.traceEnabled()) {
    kLogger.trace() << "A trace log message";
}
```

ronso0 commented:
Thanks for the hints. I'll continue to work on this soon.
Not sure about 2.3 release schedule, but I'd prioritize this if you think we already need this in the beta.

ronso0 commented:
Considering the spare time I have atm I'll roll back my direct-focus attempt and re-implement the simple table focus solution, so this improvement can go to 2.3 beta

ronso0 commented:
Done, ready to merge.

uklotzde commented on src/library/librarycontrol.cpp:
```diff
@@ -404,6 +415,12 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
 }
 
+void LibraryControl::refocusLibrary() {
+    if (!m_pLibraryWidget->hasFocus() && !m_pSidebarWidget->hasFocus()) {
```
While checking the code it seems that both m_pLibraryWidget and m_pSidebarWidget might be null, we need to check for this.

Could the check of m_pSidebarWidget be moved into setLibraryFocus(), i.e. by replacing `if (m_pSidebarWidget) {` with `(m_pSidebarWidget && !m_pSidebarWidget->hasFocus()) {` there (see below)? Explicitly focusing the sidebar widget in setLibraryFocus() if it already has the focus doesn't make sense to me.

Then this condition could be simplified to `if (m_pLibraryWidget && !m_pLibraryWidget->hasFocus()) {`.

ronso0 replied:
Thx, for the hint, ooks like you're right.
However, as far as I investigated, the functions that make use of the original `LibraryControl::setLibraryFocus()` do some existence/focus tests themselves. Maybe some of them can be consolidated.

ronso0 replied:
it's `LibraryControl::emitKeyEvent(QKeyEvent&& event)`
and `LibraryControl::slotGoToItem(double v)`

ronso0 replied:
I'm also in favor of renaming `setLibraryFocus` to `focusTracksTable`. Apart from track tables no library features' (Playlists, Crates, Computer) top pages can't be focused anyway

ronso0 commented:
Now that I dig a bit deeper I think `void LibraryControl::emitKeyEvent(QKeyEvent&& event)` should actually take care of re-focusing the tracks table:
https://github.com/mixxxdj/mixxx/blob/master/src/library/librarycontrol.cpp#L401
```
void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
    // Ensure a valid widget has the keyboard focus
    auto focusWidget = QApplication::focusWidget();
    if (!focusWidget) {
        setLibraryFocus();
        focusWidget = QApplication::focusWidget();
        if (!focusWidget) {
            return;
        }
    }
    // Send the event pointer to the currently focused widget
    for (auto i = 0; i < event.count(); ++i) {
        QApplication::sendEvent(focusWidget, &event);
    }
}
```
I assume it returns a valid `focusWidget` when library is not focused (i.e. I could alter numbers in the beatsize spinboxes with my Trax knob), thus it doesn't call set LibraryFocus() and my fix was neccesary.

That being said, I think it's more reasonable to fix `LibraryControl::emitKeyEvent()` than implementing my workaround. (apart from the improvements you suggested above)
@uklotzde What do you think?

ronso0 commented:
Should have noticed this earlier..

ronso0 commented:
Okay, got it: the easiest solution was to fix emitKeyEvent() like this (tested, works!):
```
void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
    // Ensure a valid library widget has the keyboard focus
    if (m_pLibraryWidget && !m_pLibraryWidget->hasFocus() &&
            m_pSidebarWidget && !m_pSidebarWidget->hasFocus()) {
        setLibraryFocus();
    }
    auto focusWidget = QApplication::focusWidget();

// is this check necessary after the above conditions? I'd say no..
    if (!focusWidget) {
        return;
    }
    // Send the event pointer to the currently focused widget
    for (auto i = 0; i < event.count(); ++i) {
        QApplication::sendEvent(focusWidget, &event);
    }
}
```

I'm wondering if I should apply the fix iin another PR and close this one (to keep it for documentation purpose) or if I apply the fix here and rebase.

ronso0 committed:
fix widget focus detection in LibraryControl::emitKeyEvent()

to make sure all library navigation events are targeted at the
library table when neither the treeview nor the tracks table currently
have keyboard focus.

ronso0 commented:
fixed & rebased.

ronso0 commented on src/library/librarycontrol.cpp:
```diff
@@ -389,14 +388,16 @@ void LibraryControl::slotMoveFocus(double v) {
+        setLibraryFocus();
+    }
     auto focusWidget = QApplication::focusWidget();
     if (!focusWidget) {
```
I wonder if this check is still necessary after we've tested for library & sidebar widget above?

uklotzde replied:
The invocation context might change in the future.. Use a separate check for each pointer, just to be sure:

```
VERIFY_OR_DEBUG_ASSERT(...) {
    return;
}
```

uklotzde commented on src/library/librarycontrol.cpp:
```diff
@@ -405,12 +406,10 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
-        QApplication::sendEvent(m_pSidebarWidget, &event);
-    }
+    // XXX: Set the focus of the library panel directly instead of sending tab from sidebar
+    m_pSidebarWidget->setFocus();
```
```
VERIFY_OR_DEBUG_ASSERT(m_pSidebarWidget) {
    return;
}
```

ronso0 committed:
improve checks for Library and Sidebar widgets

ronso0 commented on src/library/librarycontrol.cpp:
```diff
@@ -389,14 +388,18 @@ void LibraryControl::slotMoveFocus(double v) {
+    // Ensure a valid library widget has the keyboard focus.
+    // QApplication::focusWidget() is not sufficient here because it
+    // would return any focussed widget like WOverview, WWaveform, QSpinBox
+    VERIFY_OR_DEBUG_ASSERT(m_pSidebarWidget && m_pLibraryWidget) {
```
like this?

uklotzde replied:
Don't use && in assert statements. Otherwise you will not know which operand caused the violation. Please use separate checks as I suggested.

ronso0 replied:
alright. I read it as 'separate checks in different places' like for m_pSidebarwidget above..

ronso0 committed:
separate debug checks for Library/Sidebar widget

ronso0 commented:
I addressed the comments. Ready to merge?

uklotzde commented:
LGTM.

If any of the debug assertions should trigger we need to replace them with a conditional. But it won't hurt unless you explicitly build with debug_assertions_fatal=1.
